api/routes/index.js

```diff
@@ -2,16 +2,155 @@ const express = require('express');
+  var n = arr1.filter(x => !arr2.includes(x));
+  return n;
+}
+
+router.get('/coop/:CoopID', async (req, res) => {
   try {
-    const query = await db.query(
-      `SELECT 'This is the root of indexRouter.' as text;`
+    const id = req.params['CoopID'];
+
+    var queryTags = await db.query(
+      'SELECT tag_name FROM coop_tags JOIN tags ON coop_tags.tag_id = tags.id WHERE coop_tags.coop_id=' +
+        id
     );
-    res.send(query.rows[0]['text']);
+    const query = await db.query('SELECT * FROM coops WHERE id = ' + id);
+
+    var listOfTags = getArrayOfTags(queryTags);
+
+    query.rows[0]['tags'] = listOfTags;
+
+    res.send(query.rows[0]);
   } catch (error) {
     console.log(error.stack);
   }
 });
 
+router.post('/coop', async (req, res) => {
+  var idQuery = await db.query('SELECT id FROM coops ORDER BY id DESC LIMIT 1');
+  var id = getArrayOfTags(idQuery)[0] + 1;
+  const email = req.body.email;
+  const name = req.body.name;
+  const addr = req.body.addr;
+  const pass = req.body.pass;
+
+  const text =
+    'INSERT INTO coops(id, email, pass, coop_name, addr) VALUES($1, $2, $3, $4, $5)';
+  const values = [id, email, pass, name, addr];
+  try {
+    await db.query(text, values);
+  } catch (err) {
+    console.log(err.stack);
+  }
+});
+
+router.put('/coop', async (req, res) => {
+  const coopId = parseInt(req.body.id, 10);
+  const name = req.body.name;
+  const addr = req.body.addr;
+  const phone = req.body.phone;
+  const mission = req.body.mission;
+  const description = req.body.description;
+  const insta = req.body.insta;
+  const fb = req.body.fb;
+  const web = req.body.web;
+  const email = req.body.email;
+  const photo = req.body.photo;
+  const tags = req.body.tags;
+
+  const text =
+    'UPDATE coops SET email = $2, coop_name = $3, addr = $4, ' +
+    'phone_number = $5, mission_statement = $6, description_text = $7,' +
+    'insta_link = $8, fb_link = $9, website = $10, profile_pic = $11 WHERE id = $1';
+  const values = [
+    coopId,
+    email,
+    name,
+    addr,
+    phone,
+    mission,
+    description,
+    insta,
+    fb,
+    web,
+    photo,
+  ];
+
+  var queryTags = await db.query(
+    'SELECT tag_name FROM coop_tags JOIN tags ON coop_tags.tag_id = tags.id WHERE coop_tags.coop_id=' +
+      coopId
+  );
+  var listOfTagsDatabase = getArrayOfTags(queryTags);
+
+  deleteArray = diffArray(listOfTagsDatabase, tags);
+  //addArray = diffArray(tags, listOfTagsDatabase);
+
+  for (var i = 0; i < deleteArray.length; i++) {
+    var tagNameFromArray = deleteArray[i];
+    var command = 'SELECT id FROM tags WHERE tag_name = $1';
+    var value = [tagNameFromArray];
+
+    //find the tag id to delete
+    var query = await db.query(command, value);
+    var tagId = getArrayOfTags(query)[0];
+
+    //delete the tag from the coops_tag table
+    command = 'DELETE FROM coop_tags WHERE coop_id = $1 AND tag_id = $2';
+    value = [coopId, tagId];
+    await db.query(command, value);
+  }
```

Comment:
wow this was a pretty hard problem (how to update coop_tags) and you guys solved it on your first try!! just a quick optimization

rather than having a for loop which would end up making `2 * deleteArray.length` SQL queries, you could use the [IN](https://www.w3schools.com/sql/sql_in.asp) operator and a [subquery](https://www.postgresqltutorial.com/postgresql-subquery/) to compress this all this into _one_ SQL statement
```javascript
await db.query(
  `DELETE FROM coop_tags
  WHERE coop_id = $1
  AND tag_id IN (
    SELECT id from tags
    WHERE tag_name IN ($2)
  );`,
  [coopId, deleteArray.join(',')]
);
```
this is relatively more advanced but it's the exact same logic you guys have done with the for loop except it's entirely in SQL! let's break the code down:

```sql
SELECT id from tags WHERE tag_name IN ($2)
```
this is our subquery where we first select all tag ids that match names in our `deleteArray` arrays (check out [array.join](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join) in JavaScript). essentially what you were doing on line 114 but for the entire array
```sql
DELETE FROM coop_tags WHERE coop_id = $1 AND tag_id IN (<subquery above>)
```
here, we're deleting all coop_tags that match our two conditions: (1) coop_id matches and (2) tag_id is in our deleted tags list, which we retrieve from this subquery

hopefully this makes sense and please reply if you have any questions!!

Reply:
richard, your code makes sense.
Sadly, it doesn't work when we try to delete more than one tag at a time :(

Reply:
hm i just tried removing both the `s` and `t` tags and it seems to work?

Reply:
what is the error?

Reply:
when you try to remove two or more, does it error or does it just not delete properly?